skbio/alignment/_tabular_msa.py

```diff
@@ -716,8 +730,119 @@ def reindex(self, key=None, keys=None):
             keys_.flags.writeable = False
 
         self._keys = keys_
+        self._minter = minter
+
+    def _fail_if_duplicate_keys(self, keys):
+        # Hashability of keys is implicitly checked here.
+        duplicates = find_duplicates(keys)
+        if duplicates:
+            raise UniqueError(
+                "Keys must be unique. Duplicate keys: %r" % duplicates)
 
     @experimental(as_of='0.4.0-dev')
+    def append(self, sequence, key=None, minter=None):
+        """Append a sequence to the MSA.
+
+        Parameters
+        ----------
+        sequence : alphabet-aware scikit-bio sequence object
+            Sequence to be appended. Must match the dtype of the MSA and the
+            number of positions in the MSA.
+        minter : callable or metadata key, optional
+            If None, the MSA is checked to see if a previous minter has been
+            cached. If not, an exception is raised. If a cached key is found
+            it is used to reindex the MSA.
+
+        Raises
+        ------
+        TypeError
+            If the sequence object is a type that doesn't have an alphabet
+        TypeError
+            If the type of the sequence does not match the dtype of the MSA.
+        ValueError
+            If the length of the sequence does not match the number of
+            positions in the MSA.
+        OperationError
+            If no key is provided, but the MSA has keys, and no cached key
+            is found.
+
+        See Also
+        --------
+        reindex
+
+        Notes
+        -----
+        The MSA is not automatically re-aligned when a sequence is appended.
+        Therefore, this operation is not necessarily meaningful on its own.
+
+        Examples
+        --------
+        >>> from skbio import DNA, TabularMSA
+        >>> msa = TabularMSA([DNA('')])
+        >>> msa.append(DNA(''))
+        >>> msa == TabularMSA([DNA(''), DNA('')])
+        True
+
+        >>> msa = TabularMSA([DNA('', metadata={'id': 'a'})], minter='id')
+        >>> msa.append(DNA('', metadata={'id': 'b'}), minter='id')
+        >>> msa == TabularMSA([DNA('', metadata={'id': 'a'}),
+        ...                    DNA('', metadata={'id': 'b'})], minter='id')
+        True
+        """
+
+        if key is not None and minter is not None:
```

Comment:
Not sure if this will help simplify the code, but can you handle the case where the key has been successfully added but appending the sequence fails? In this case we'd want the MSA's state to remain consistent -- both the key *and* the sequence must succeed otherwise no changes are made to MSA state. Can you also add a unit test for this behavior?